test/TesterInternal/GeoClusterTests/LogViewProviderTestBattery.cs

```diff
@@ -0,0 +1,490 @@
+                }
+
+                // push & get all
+                AssertEqual(numupdates, Client[0].GetAGlobal(grainClass, x), grainidentity); 
+
+                for (int i = 1; i < numclusters; i++)
+                    AssertEqual(numupdates, Client[i].GetAGlobal(grainClass, x), grainidentity); // get all
+
+                // check versions
+                for (int i = 0; i < numclusters; i++)
+                    AssertEqual(numupdates, Client[i].GetConfirmedVersion(grainClass, x), grainidentity);
+            });
+
+            Func<Task> checker4 = () => Task.Run(() =>
+            {
+                int x = GetRandom();
+                var t = new List<Task>();
+                for (int i = 0; i < numclusters; i++)
+                {
+                    int c = i;
+                    t.Add(Task.Run(() => Assert.True(Client[c].GetALocal(grainClass, x) == 0)));
+                }
+                for (int i = 0; i < numclusters; i++)
+                {
+                    int c = i;
+                    t.Add(Task.Run(() => Assert.True(Client[c].GetAGlobal(grainClass, x) == 0)));
+                }
+                return Task.WhenAll(t);
+            });
+
+            Func<Task> checker5 = () => Task.Run(() =>
+            {
+                var x = GetRandom();
+                Task.WaitAll(
+                   Task.Run(() =>
+                   {
+                       Client[0].AddReservationLocal(grainClass, x, 0);
+                       Client[0].RemoveReservationLocal(grainClass, x, 0);
+                       Client[0].Synchronize(grainClass, x);
+                   }),
+                 Task.Run(() =>
+                 {
+                     Client[1].AddReservationLocal(grainClass, x, 1);
+                     Client[1].RemoveReservationLocal(grainClass, x, 1);
+                     Client[1].AddReservationLocal(grainClass, x, 2);
+                     Client[1].Synchronize(grainClass, x);
+                 })
+               );
+                var result = Client[0].GetReservationsGlobal(grainClass, x);
+                Assert.Equal(1, result.Length);
+                Assert.Equal(2, result[0]);
+            });
+
+            Func<int, Task> checker6 = async (int preload) =>
+            {
+                var x = GetRandom();
+
+                if (preload % 2 == 0)
+                    Client[1].GetAGlobal(grainClass, x);
+                if ((preload / 2) % 2 == 0)
+                    Client[0].GetAGlobal(grainClass, x);
+
+                bool[] done = new bool[numclusters - 1];
+
+                var t = new List<Task>();
+
+                // create listener tasks
+                for (int i = 1; i < numclusters; i++)
+                {
+                    int c = i;
+                    t.Add(Task.Run(() =>
+                    {
+                        while (Client[c].GetALocal(grainClass, x) != 1)
+                            System.Threading.Thread.Sleep(100);
```

Comment:
The clients are in a different AppDomain so they are called synchronously.

But I suppose the sleep can be replaced with Task.Delay. I will do that.